Die neue Spring-Doku (#87)
it-ninjas/labs: content/en/docs/java/new-spring-doc.md

        return ResponseEntity.ok().build();
    }
}
```

Dies ist eine Rest-Resource und sie wird definiert mit der Annotation `@RestController`. Die
Annotation `@RequestMapping("/orders")` legt fest, dass alle Aufrufe, wessen URL mit `"/orders"` beginnen, diesen Rest-Controller
verwenden sollen.

Schauen wir uns dies doch ein bisschen genauer an:

```java
@GetMapping("/{id}")
public Order findById(@PathVariable Long id){
    return orderService.findById(id).orElseThrow(EntityNotFoundException::new);
}
```

Mit der `@GetMapping` Annotation bestimmen wir, dass alle Anfragen auf dem `orders/id` Pfad von der
annotierten Methode gehandelt werden, wenn die HTTP-Methode `GET` verwendet wurde. `@PathVariable` bestimmt, dass die ID als Path-Variable (also in der URL)
angegeben wird. Das bedeutet, dass ein Aufruf auf `/orders/12` das gleiche Ergebnis hat wie `findByID(12)`.

```java
@PutMapping("/{id}")
public Order update(@PathVariable Long id,@RequestBody Order order){
    return orderService.update(id,order);
}
```

Die Annotation `@RequestBody` wird verwendet, um anzugeben, dass der Parameter `order` aus dem Request-Body des 
HTTP-Requests gelesen werden soll.

#### Control Layer

> Der Control Layer bildet den Kern aller Anwendungen und enthält dessen Geschäfts-Logiken. Auf der
> technischen Ebene ist der Control Layer die grundlegendste und die am wenigsten interessante Schicht.

Reviewer: ```suggestion
Der Control Layer bildet den Kern aller Anwendungen und enthält dessen Geschäfts-Logiken. Auf der
technischen Ebene ist der Control Layer die grundlegendste Schicht.
```

Die Control Layer könnte wie folgt aussehen:

```java
@Service
@Transactional
public class OrderService {
    private final OrderRepository orderRepository;

    public OrderService(OrderRepository orderRepository) {
        this.orderRepository = orderRepository;
    }

    public List<Order> findAll() {
        return (List<Order>) orderRepository.findAll();
    }

    public Optional<Order> findById(Long id) {
        return orderRepository.findById(id);
    }

    public Order save(Order order) {
        return orderRepository.save(order);
    }

    public Order update(Long id, Order order) {
        Order toUpdate = orderRepository.findById(id).orElseThrow(EntityNotFoundException::new);
        toUpdate.setName(order.getName());
        return toUpdate;
    }

    public void delete(Long id) {
        orderRepository.deleteById(id);
    }
}
```

Diese Klasse besitzt nur zwei Annotationen nämlich `@Service` und
`@Transactional`. Die `@Service` Annotation markiert Beans als Holder
der Businesslogik. `@Transactional` sagt einfach, dass alle Funktionen in der Klasse in einer

> einzigen Transaktion ausgeführt werden sollen. Das musst du vorläufig aber noch nicht verstehen.

Reviewer: ```suggestion
einzigen Transaktion ausgeführt werden sollen. Das musst du vorläufig noch nicht zu 100% verstehen.
```

Profil deaktiviert.
Hier ist ein Beispiel:

```java
@Component
@Profile("test")
public class TestString {
    @Bean
    public String test() {
        return "test";
    }
}
```

Im Beispiel lädt die Komponente `TestString` nur, wenn das Profil `test` aktiv ist.  
Bei der `@Profile`-Annotation kann man auch logische Operatoren wie "Und" (`&`) , "Oder" (`|`) und "Nicht" (`!`) benutzen.
Hier ein Beispiel:

```java
@Component
public class Demo {
    @Bean
    @Profile("default")
    public String defaultString() {
        return "Standard Profil";
    }
    
    @Bean
    @Profile("test | test2")
    public String testString() {
        return "Test Test";
    }
    
    @Bean
    @Profile("!(default | test | test2)")
    public String rewoltString() {
        return "None of the above";
    }
}
```

In diesem Beispiel werden die Methoden angesteuert. Wenn das Profil `default` aktiv ist, wird
die `defaultString` Methode geladen, wenn das Profil `test` oder `test2` aktiv ist, wird
die `testString` Methode geladen, wenn keines der obengenannten Profile aktiv ist, wird
die `rewoltString`-Methode geladen.

> 

Reviewer: Sugesstion für den oberen Kommentar:
```suggestion
> Hinweis: Wenn eine Methode oder andere Funktionalität im aktuellen Profil deaktiviert ist, so werden jegliche Aufrufe von Spring automatisch ignoriert. Es wird also keine Compiler- oder RunTime-Exception geben.
```

Spring genutzt werden.

### Inversion of Control (IoC)
Inversion of Control (IoC) ist ein Prinzip, das vereinfacht gesagt den Kontroll-Fluss umkehrt. Das Ziel dabei ist es, eine lose Kopplung zu erreichen, wodurch der Code einfacher testbar, wartbarer und erweiterbarer wird.

IoC wird in Spring durch "Dependency Injection" erreicht, was im nächsten Kapitel vorgestellt wird. Dabei werden Abhängigkeiten wie Variablen von einer anderen Klasse durch Spring automatisch erstellt und "injektiert" (eingefügt), statt dass man sich selbst darum kümmern und Objekte selbst erstellen muss.
Zuständigkeiten getrennt werden. Spring nutzt dieses Prinzip.

### Dependency Injection (DI)
DI bezeichnet die Weitergabe einer Abhängigkeit an ein Objekt. Spring verwendet einen Inversion-of-Control-Container, 
um zu bestimmen, wie und wo die Injektion erfolgt. Es gibt verschiedene Orte im Code, wo DI eingesetzt werden kann. Es gibt die Konstruktor-, 
Setter- und Feld-basierte Injection.

##### Konstruktor Injection

Bei einer Konstruktor-Injection fungiert jedes Argument als eine Abhängigkeit.  
Ein Beispiel sieht so aus:

```java

> @Configuration
> public class AppConfig {
> 
>   @Bean
>   public Item item1() {
>     return new ItemImpl1();
>   }
> 
>   @Bean
>   public Store store() {
>     return new Store(item1());
>   }
> }

Author: Damit es wirklich Konstruktor-Injection ist, müsste der `Store` noch eigenständig definiert werden:

```suggestion
@Component
public class Store {
    private final Item item;

    public Store(Item item1) {
        this.item = item1;
    }
}
```

```

> Die `@Configuration` Annotation besagt das diese Klasse die Konfiguration beinhalten

Author: ```suggestion
In diesem Beispiel wird die Bean `item1` injektiert.
```

- **XML-Konfiguration:** Historisch gesehen wurden Beans oft in XML-Dateien definiert. Hier wird beschrieben, wie der Container diese Instanzen erstellt und verwaltet.

  ```xml
  <bean id="exampleBean" class="com.example.ExampleClass">
      <property name="someProperty" value="Wert"/>
  </bean>
  ```

- **Java-Konfiguration:** Mit Annotationen und Java-Klassen kann die Konfiguration von Beans lesbarer und flexibler gestaltet werden.

  ```java
  @Configuration
  public class AppConfig {
      @Bean
      public ExampleClass exampleBean() {
          return new ExampleClass();
      }
  }
  ```

#### Bean-Scopes

Beans haben verschiedene Scopes, die ihr Lebenszyklus bestimmen:

- **Singleton:** Es wird nur eine Instanz pro Container erzeugt. Dies ist der Standard-Scope in Spring.
- **Prototype:** Für jeden Aufruf zur Bean-Anforderung wird eine neue Instanz erstellt.
- **Request, Session, Application:** Spezifische Scopes für Webanwendungen, um den Lebenszyklus der Beans an HTTP-Anforderungen oder Sitzungen zu binden.

#### Verwendung von Beans

Beans werden normalerweise über Injektionen verwendet, um Abhängigkeiten zwischen verschiedenen Komponenten einer 

> Anwendung zu verwalten. Dies geschieht über Annotationen wie `@Autowired` oder `@Inject`:

Author: ```suggestion
Anwendung zu verwalten. Dies geschieht via Dependency Injection:
```

```java
@Component
public class ExampleService {
    private final ExampleRepository repository;

    public ExampleService(ExampleRepository repository) {
        this.repository = repository;
    }

    // ...
}
```

> Hier wird `ExampleRepository` als Abhängigkeit von `ExampleService` injiziert.

Author: ```suggestion
Hier wird `ExampleRepository` als Abhängigkeit von `ExampleService` injiziert. Die `@Component`-Annotation wird für dieses Beispiel verwendet, damit die Abhängigkeiten via Dependency Injection eingefügt werden.
```

##### Scope

Der Scope eines Beans besagt, wann und wie ein Bean erstellt wird.

Hier eine Liste der Scopes:

| Scope       | Beschreibung                                                                                   |
|-------------|------------------------------------------------------------------------------------------------|
| singleton (Standard) | Singleton besagt, dass es immer nur eine Instanz eines Beans gibt, welche dann geteilt wird. Es wird daher nur **eine** Bean-Instanz in jedem IoC-Container.  Mehr [hier](#singleton). |
| prototype   | Erstellt bei jeder Abhängigkeit eine neue Bean-Instanzen.                                            |
| request     | Erstellt für jeden HTTP Request eine Bean Instanz.                                             |
| session     | Erstellt für jede HTTP `Session` eine Bean Instanz.                                            |
| application | Erstellt für jeden `ServetContext` eine Bean Instanz.                                          |
| websocket   | Erstellt für jeden `WebSocket` eine Bean Instanz.                                              |

Die beiden wichtigsten Scopes sind `singleton` und `prototype`.

> Singleton besagt, dass es immer nur eine Instanz eines Beans gibt, welche dann geteilt wird. Mehr
> dazu [hier](#singleton).
> 

Author: ```suggestion
```

> Prototype besagt, dass für jede Abhängigkeit ein neues Bean erstellt wird.  
> Hier ein Bild:
> 
> ![PrototypeScopeBild](../new-spring-doc/prototype.png)
> 
> Quelle: https://docs.spring.io/spring-framework/docs/current/reference/html/images/prototype.png

Author: ```suggestion
```

Zuständigkeiten getrennt werden. Spring nutzt dieses Prinzip.

### Dependency Injection (DI)
DI bezeichnet die Weitergabe einer Abhängigkeit an ein Objekt. Spring verwendet einen Inversion-of-Control-Container, 
um zu bestimmen, wie und wo die Injektion erfolgt. Es gibt verschiedene Orte im Code, wo DI eingesetzt werden kann. Es gibt die Konstruktor-, 
Setter- und Feld-basierte Injection.

##### Konstruktor Injection

Bei einer Konstruktor-Injection fungiert jedes Argument als eine Abhängigkeit.  
Ein Beispiel sieht so aus:

```java

@Configuration
public class AppConfig {

  @Bean
  public Item item1() {
    return new ItemImpl1();
  }

  @Bean
  public Store store() {
    return new Store(item1());
  }
}
```

Die `@Configuration` Annotation besagt das diese Klasse die Konfiguration beinhalten

Die `@Bean` Annotation besagt, dass es sich hier um ein Bean handelt. Mehr dazu [hier](#bean).

> ##### Setter Injection
> 
> Bei der Setter basierten Injection erstellt man Konstruktoren ohne Argumente. Die Injection passiert
> dann mit dem Setter. Es wird empfohlen, obligatorische Abhängigkeiten mit Konstruktor Injections zu
> machen während man optionale Abhängigkeiten mit Setter Injections macht.  
> Hier ein Beispiel:
> 
> ```java
> @Bean
> public Store store() {
>     Store store=new Store();
>     store.setItem(item1());
>     return store;
> }
> ```

Author: ```suggestion
```

können Singleton-Beans erstellt werden, aber es ist wichtig zu beachten, dass dies innerhalb des Spring-Containers 
gilt, nicht systemweit.

### Beans

Beans sind Kernkomponenten in der Spring-Welt. Sie repräsentieren die grundlegenden Bausteine, die vom Spring 
IoC-Container verwaltet werden. Diese Container verwalten die Lebenszyklen der Beans, wodurch ihre Erstellung, 
Initialisierung, Verwendung und Entsorgung standardisiert werden.

#### Erstellung von Beans

Beans können auf verschiedene Weisen erstellt werden:

- **XML-Konfiguration:** Historisch gesehen wurden Beans oft in XML-Dateien definiert. Hier wird beschrieben, wie der Container diese Instanzen erstellt und verwaltet.

  ```xml
  <bean id="exampleBean" class="com.example.ExampleClass">
      <property name="someProperty" value="Wert"/>
  </bean>
  ```

- **Java-Konfiguration:** Mit Annotationen und Java-Klassen kann die Konfiguration von Beans lesbarer und flexibler gestaltet werden.

  ```java
  @Configuration
  public class AppConfig {
      @Bean
      public ExampleClass exampleBean() {
          return new ExampleClass();
      }
  }
  ```

> #### Bean-Scopes
> 
> Beans haben verschiedene Scopes, die ihr Lebenszyklus bestimmen:
> 
> - **Singleton:** Es wird nur eine Instanz pro Container erzeugt. Dies ist der Standard-Scope in Spring.
> - **Prototype:** Für jeden Aufruf zur Bean-Anforderung wird eine neue Instanz erstellt.
> - **Request, Session, Application:** Spezifische Scopes für Webanwendungen, um den Lebenszyklus der Beans an HTTP-Anforderungen oder Sitzungen zu binden.

Author: ```suggestion
```

#### Verwendung von Beans

Beans werden normalerweise über Injektionen verwendet, um Abhängigkeiten zwischen verschiedenen Komponenten einer 
Anwendung zu verwalten. Dies geschieht über Annotationen wie `@Autowired` oder `@Inject`:

```java
@Component
public class ExampleService {
    private final ExampleRepository repository;

    @Autowired
    public ExampleService(ExampleRepository repository) {
        this.repository = repository;
    }

    // ...
}
```

Hier wird `ExampleRepository` als Abhängigkeit von `ExampleService` injiziert.

Beans sind flexibel, wiederverwendbar und ermöglichen die Modularität von Anwendungen. Sie bilden das Rückgrat vieler 
Spring-Anwendungen und spielen eine entscheidende Rolle in der Komponentenverwaltung und -integration.

> [Hier ist die offizielle Bean
> Dokumentation](https://docs.spring.io/spring-framework/docs/current/reference/html/core.html#beans-definition)  
> [Hier ist die offizielle `@Bean` Annotation Dokumentation](https://docs.spring.io/spring-framework/docs/current/reference/html/core.html#beans-java-bean-annotation)

Author: ```suggestion
In der folgenden offiziellen Dokumentation zu der `@Bean`-Annotation findest du weitere Beispiele, wie Beans verwendet werden:
[Using the Bean Annotation](https://docs.spring.io/spring-framework/docs/current/reference/html/core.html#beans-java-bean-annotation)
```

##### Scope

Der Scope eines Beans besagt, wann und wie ein Bean erstellt wird.

Hier eine Liste der Scopes:

| Scope       | Beschreibung                                                                                   |
|-------------|------------------------------------------------------------------------------------------------|

> | singleton   | (Standard) Erstellt nur **eine** Bean Instanz in jedem IoC container. Mehr [hier](#singleton). |

Author: ```suggestion
| singleton   | (Standard) Singleton besagt, dass es immer nur eine Instanz eines Beans gibt, welche dann geteilt wird. Es wird daher nur **eine** Bean-Instanz in jedem IoC-Container.  Mehr [hier](#singleton). |
```

> | prototype   | Erstellt eine unbegrenzte Anzahl an Bean Instanzen.                                            |

Author: ```suggestion
| prototype   | Erstellt bei jeder Abhängigkeit eine neue Bean-Instanzen.                                            |
```

um zu bestimmen, wie und wo die Injektion erfolgt. Es gibt verschiedene Orte im Code, wo DI eingesetzt werden kann. Es gibt die Konstruktor-, 
Setter- und Feld-basierte Injection.

##### Konstruktor Injection

Bei einer Konstruktor-Injection fungiert jedes Argument als eine Abhängigkeit.  
Ein Beispiel sieht so aus:

```java

@Configuration
public class AppConfig {

  @Bean
  public Item item1() {
    return new ItemImpl1();
  }

  @Bean
  public Store store() {
    return new Store(item1());
  }
}
```

Die `@Configuration` Annotation besagt das diese Klasse die Konfiguration beinhalten

Die `@Bean` Annotation besagt, dass es sich hier um ein Bean handelt. Mehr dazu [hier](#bean).

##### Setter Injection

Bei der Setter basierten Injection erstellt man Konstruktoren ohne Argumente. Die Injection passiert
dann mit dem Setter. Es wird empfohlen, obligatorische Abhängigkeiten mit Konstruktor Injections zu
machen während man optionale Abhängigkeiten mit Setter Injections macht.  
Hier ein Beispiel:

```java
@Bean
public Store store() {
    Store store=new Store();
    store.setItem(item1());
    return store;
}
```

##### Felder Injection

> Bei der Felder Injection werden die Abhängigkeiten direkt in die Felder Injected. Es ist jedoch nicht
> empfehlenswert diese Art der injection zu benutzen, da diese Reflection benutzt, was weniger Effizient
> ist.  

Author: ```suggestion
Bei der Felder Injection werden die Abhängigkeiten direkt in die Felder Injected.
```

Hier ein Beispiel:

```java
public class Store {
    @Autowired
    private Item item;
}
```

> 

Author: ```suggestion

###### Feld- vs. Konstruktor-Injection

* Immutable-Abhängigkeiten: Bei der Konstruktor-Injection ist es möglich, alle Variablen, die im Konstruktor gesetzt werden, mit `final` zu deklarieren.
* Erforderliche Abhängigkeiten: Bei der Konstruktor-Injection sind alle Abhängigkeiten per Default erforderlich, während sie bei der Field-Injection standardmässig optional sind.

```

> ##### Automatische Abhängigkeiten Injection (Autowired)
> 
> Spring kann Abhängigkeiten auch automatisch erkennen und diese einsetzen. Hierbei git es vier verschiedenen
> Modi.
> 
> * `no` Standartwert, hier wird kein autowiring gemacht.
> * `byName` Hier werden die Abhängigkeiten dem Namen nach eingesetzt.
> * `byType` Hier werden die Abhängigkeiten dem Typ nach eingesetzt. Gibt es mehrere Beans eines Typus,
>   dann gibt es einen Fehler.
> * `consturctor` Hier werden die Abhängigkeiten der Konstruktoren Argumente nach eingesetzt. Das
>   bedeutet, dass Spring nach Beans sucht, welche denselben Typ wie die Konstruktoren Argumente haben.
> 
> Hier ist ein Beispiel mit `byType`:
> 
> ```java
> 
> @Bean(autowire = Autowire.BY_TYPE)
> public class Store {
>     private Item item;
>     
>     public setItem(Item item) {
>         this.item = item;
>     }
> }
> ```
> 
> Mit der `@Aurowired` Annotation können wir auch mit Type einsetzten:
> 
> ```java
> public class Store {
>     @Autowired
>     private Item item;
> }
> ```
> 
> Wenn es mehr als eine Bean des gleichen typen gibt, dann kann man mit `@Qualifier` einen Namen
> bestimmen:
> 
> ```java
> public class Store {
>     @Autowired
>     @Qualifier("item1")
>     private Item item;
> }
> ```
> 
> Quelle und weitere Informationen: https://www.baeldung.com/inversion-control-and-dependency-injection-in-spring

Author: ```suggestion
```

### Singleton

Das Singleton Design Pattern stellt sicher, dass von einer Klasse nur eine Instanz existiert. Im Kontext von Spring 
können Singleton-Beans erstellt werden, aber es ist wichtig zu beachten, dass dies innerhalb des Spring-Containers 

> gilt, nicht systemweit. Hier ein Beispielbild:

Author: ```suggestion
gilt, nicht systemweit.
```

> 
> ![beispielBild](https://www.baeldung.com/wp-content/uploads/2023/02/singleton_design_pattern3.png)
> 
> Quelle: https://www.baeldung.com/wp-content/uploads/2023/02/singleton_design_pattern3.png

Author: ```suggestion
```

> ### HTTP Access-Control
> 
> Das HTTP-Protokoll bietet folgende "HTTP-Methoden" an, um auf sogenannte Ressourcen ("Endpunkte" bzw. URLs) zuzugreifen:
> 
> * GET – Ressourcen erhalten
> * POST – Neue ressourcen erstellen
> * DELETE – Ressource löschen
> * PUT – Ressourcen aktualisieren oder gegebenenfalls neu erstellen

Author: ```suggestion
```

Datenbank.

Hier ist eine beispiel Konfiguration mit einer Mariadb Datenbank:

```properties
spring.datasource.url=jdbc:mariadb://localhost:3306/order
spring.datasource.username=spring
spring.datasource.password=1234
spring.datasource.driver-class-name=org.mariadb.jdbc.Driver
spring.jpa.hibernate.ddl-auto=update
```

### Profile

In Spring kann man verschiedene Profile erstellen. Damit kann man sicherstellen, dass bestimmte `@Bean`s
nur entsprechend gesetzt werden, wenn eine bestimmte Bedingung zutrifft. Die Datenbank soll zum Beispiel nur Beispieldaten
laden, wenn das `Dev`-Profil aktiv ist, ansonsten soll eine richtige Datenbank verwendet werden.

Mit der `@Profile` Annotation kann man einer Klasse oder Methode (Beans) sagen, ob sie bei einem
Profil läuft. Das standard Profil ist `default`, wenn irgendein Profil aktiv ist, wird das `default`
Profil deaktiviert.
Hier ist ein Beispiel:

```java
@Component
@Profile("test")
public class TestString {
    @Bean
    public String test() {
        return "test";
    }
}
```

Im Beispiel lädt die Komponente `TestString` nur, wenn das Profil `test` aktiv ist.  

> Bei der `@Profile` Annotation kann man auch logische operatoren wie nicht (`!`), und (`&`) und
> oder (`|`) benutzen.  

Reviewer: ```suggestion
Bei der `@Profile` Annotation kann man auch logische operatoren wie und (`&`) , oder (`|`) und
nicht (`!`) benutzen.  
```

---
title: "New Spring Doc"
linkTitle: "New Spring Doc"
weight: 16
description: >
  Modul #J10 - Spring Framework
---

#### Ziele

* Ich weiss, was ein Java-Framework ist.
* Ich kenne einige Vor- und Nachteile beim Verwenden eines Java-Frameworks.
* Ich kenne ein paar Module des Spring-Frameworks (z.B. Core, Data-Access, Web usw.) und weiss,
  welche Aufgaben sie erfüllen.
* Ich kenne die wichtigsten Spring-Projekte (z.B. Spring-Boot, Spring-Cloud, Spring-Security usw.)
  und weiss wofür ich diese einsetzen kann.
* Ich weiss, was "Inversion of Control" (IoC) bedeutet und kenne dessen Vorteile.
* Ich weiss, was "Dependency-Injection" (DI) ist und kenne die verschiedene Möglichkeiten (z.B.
  Constructor-Based, Setter-Based, Field-Based) um DI in Spring zu benutzen.
* Ich weiss, was ein "Spring-Bean" ist und wie es in Spring erzeugt und verwendet wird.
* Ich kenne die verschiedene Scopes eines "Spring-Beans" (z.B. Singleton, Session, Application
  usw.).
* Ich kenne die, am häufigsten verwendeten, Spring-Bean Annotationen (z.B. `@Configuration`,

>   `@Component`, `@Repository`, `@Controller` usw.), wie ich diese einsetze und welchem Zweck sie

Author: ```suggestion
  `@Component`, `@Repository`, `@Controller` und weiss, wie ich diese einsetze und welchem Zweck sie
```

  dienen.
* Ich weiss, was "Wiring" ist und wie es in Spring gemacht wird.
* Ich weiss, wofür Properties in Spring verwendet werden und wie ich Properties in Spring setzen

>   kann.

Author: ```suggestion
```

Projekts überflüssig macht. Spring Boot ist jedoch kein Ersatz für das Spring Framework, sondern vereinfacht die 
Konfiguration.

## Design-Pattern
Design-Patterns sind wiederverwendbare Vorlagen für die Anwendungsentwicklung und sind nicht spezifisch für eine 
Programmiersprache. Es gibt verschiedene Arten, z. B. Kreationsmuster, Strukturmuster und Verhaltensmuster, die auch in 
Spring genutzt werden.

### Inversion of Control (IoC)
Inversion of Control ist ein Prinzip, das besagt, dass eine Entität nur eine Hauptaufgabe ausführt, wodurch die 
Zuständigkeiten getrennt werden. Spring nutzt dieses Prinzip.

### Dependency Injection (DI)
DI bezeichnet die Weitergabe einer Abhängigkeit an ein Objekt. Spring verwendet einen Inversion-of-Control-Container, 
um zu bestimmen, wie und wo die Injektion erfolgt. Es gibt verschiedene Möglichkeiten der DI, darunter Konstruktor-, 
Setter- und Feld-basierte Injection.

##### Konstruktor Injection

Bei einer Konstruktor Injection fungiert jedes Argument als eine Abhängigkeit.  
Ein Beispiel sieht so aus:

```java

@Configuration
public class AppConfig {

  @Bean
  public Item item1() {
    return new ItemImpl1();
  }

  @Bean

>   public Store store() {

Author: Kommentar rein in die Doku.

    return new Store(item1());
  }
}
```

> Die `@Configuration` Annotation besagt das diese Klasse die Konfiguration beinhalten

Author: Können wir gerne rausnehmen.

  ```xml
  <bean id="exampleBean" class="com.example.ExampleClass">
      <property name="someProperty" value="Wert"/>
  </bean>
  ```

- **Java-Konfiguration:** Mit Annotationen und Java-Klassen kann die Konfiguration von Beans lesbarer und flexibler gestaltet werden.

  ```java
  @Configuration
  public class AppConfig {
      @Bean
      public ExampleClass exampleBean() {
          return new ExampleClass();
      }
  }
  ```

#### Bean-Scopes

Beans haben verschiedene Scopes, die ihr Lebenszyklus bestimmen:

- **Singleton:** Es wird nur eine Instanz pro Container erzeugt. Dies ist der Standard-Scope in Spring.
- **Prototype:** Für jeden Aufruf zur Bean-Anforderung wird eine neue Instanz erstellt.
- **Request, Session, Application:** Spezifische Scopes für Webanwendungen, um den Lebenszyklus der Beans an HTTP-Anforderungen oder Sitzungen zu binden.

#### Verwendung von Beans

Beans werden normalerweise über Injektionen verwendet, um Abhängigkeiten zwischen verschiedenen Komponenten einer 
Anwendung zu verwalten. Dies geschieht über Annotationen wie `@Autowired` oder `@Inject`:

```java

> @Component

Author: siehe nächste Suggestion.

tun haben definiert. Mit der `@Controller` Annotation werden Komponenten, welche mit der Kommunikation mit
der Aussenwelt zu tun haben definiert. In der `@Service` Annotation werden Komponenten, welche mit der Kontrolllogik zu
tun haben definiert.

### Repository

Eines der am häufigsten verwendeten Design-Patterns (siehe oben) ist das Repository Design-Pattern.
Es ist ein Interface, welches dafür zuständig ist, uns Zugang zu den Daten in der Datenbank zu bereiten.
Man kann es sich ein wenig, wie einen Bibliothekar vorstellen. Er gibt uns die Bücher zurück, die wir
wollen, ohne dass wir wissen, wie er sie uns beschafft. Wir kennen nur die API, also die Befehle
und welche Resultate wir erwarten können.

```java
public interface ArticleRepository {
    List<Article> readAll();
    
    List<Article> readLatest();
    
    List<Article> readByTags(Tag... tags);
    
    Article readById(long id);
    
    Article create(Article article);
    
    Article update(Article article);
    
    Article delete(Article article);
}
```

Meistens wird das Interface im Hintergrund vom Framework implementiert und wir müssen nichts dafür
tun.

### Boundary Layer

> In diesem Layer definieren wir unsere REST Ressourcen. Hier ein Beispiel einer Order-Klasse in einem

Reviewer: Wird im nächsten Commit ergänzt

    
    @PostMapping
    public Order save(@RequestBody Order order) {
        return orderService.save(order);
    }
    
    @PutMapping("/{id}")
    public Order update(@PathVariable Long id, @RequestBody Order order) {
        return orderService.update(id, order);
    }
    
    @DeleteMapping("/{id}")
    public ResponseEntity delete(@PathVariable Long id) {
        orderService.delete(id);
        return ResponseEntity.ok().build();
    }
}
```

Dies ist eine Rest-Resource und sie wird definiert mit der Annotation `@RestController`. Die
Annotation `@RequestMapping("/orders")` legt fest, dass alle Aufrufe auf Orders diese Rest-Resource
verwenden sollen.

Schauen wir uns dies doch ein bisschen genauer an:

```java
@GetMapping("/{id}")
public Order findById(@PathVariable Long id){
    return orderService.findById(id).orElseThrow(EntityNotFoundException::new);
}
```

Mit der `@GetMapping` Annotation bestimmen wir, dass alle Anfragen auf dem `orders/id` Pfad von der
annotierten Methode gehandelt werden. `@PathVariable` bestimmt, dass die ID als Path-Variable
angegeben wird. Das bedeutet, dass ein Aufruf auf `/orders/12` das gleiche Ergebnis hat wie `findByID(12)`.

```java

> @PutMapping("/{id}")

Reviewer: Sollte mit der Erklärung von REST im nächsten Commit klar sein.

`jdbc:[datenbanktyp]://[server]:[port]/[datenbank]`)

`spring.datasource.username` gibt den benutzername der Datenbank an

`spring.datasource.password` gibt das Passwort der Datenbank an

`spring.datasource.driver-class-name` gibt an wo der Datenbank treiber ist (muss als maven
Dependency hinzugefügt werden)

`spring.jpa.hibernate.ddl-auto` sagt wie Spring mit der Datenbank umgehen soll. `update` z.B. sorgt
dafür, dass fehlende Spalten oder dergleichen automatisch erstellt werden und dass nie Sachen
gelöscht werden. `create-drop` erstellt die Datenbank und löscht die Objekte nachher. `none` belässt die
Datenbank.

Hier ist eine beispiel Konfiguration mit einer Mariadb Datenbank:

```properties
spring.datasource.url=jdbc:mariadb://localhost:3306/order
spring.datasource.username=spring
spring.datasource.password=1234
spring.datasource.driver-class-name=org.mariadb.jdbc.Driver
spring.jpa.hibernate.ddl-auto=update
```

### Profile

In Spring kann man verschiedene Profile erstellen. Damit kann man sicherstellen, das bestimmte Dinge
nur dann gemacht werden, wenn man sie braucht. Die Datenbank soll zum Beispiel nur Beispieldaten

> laden, wenn das Dev Profil aktiv ist.

Author: ```suggestion
laden, wenn das `Dev`-Profil aktiv ist, ansonsten soll eine richtige Datenbank verwendet werden.
```

Mit der `@Profile` Annotation kann man einer Klasse oder Methode (Beans) sagen, ob sie bei einem
Profil läuft. Das standard Profil ist `default`, wenn irgendein Profil aktiv ist, wird das `default`
Profil deaktiviert.
Hier ist ein Beispiel:

```java
@Component
@Profile("test")
public class TestString {
    @Bean
    public String test() {
        return "test";
    }
}
```

> Im Beispiel lädt die Komponente `TestString` nur wenn das Profil `test` aktiv ist.  

Author: ```suggestion
Im Beispiel lädt die Komponente `TestString` nur, wenn das Profil `test` aktiv ist.  
```

Bei der `@Profile` Annotation kann man auch logische operatoren wie nicht (`!`), und (`&`) und
oder (`|`) benutzen.  
Hier ein Beispiel:

```java
@Component
public class Demo {
    @Bean

>     @Profile("default")

Author: Welches Verhalten treffen wir an, wenn das entsprechende Modul nicht aktiv ist?

    public String defaultString() {
        return "Standard Profil";
    }
    
    @Bean
    @Profile("test | test2")
    public String testString() {
        return "Test Test";
    }
    
    @Bean
    @Profile("!(default | test | test2)")
    public String rewoltString() {
        return "None of the above";
    }
}
```

In diesem Beispiel werden die Methoden angesteuert. Wenn das Profil `default` aktiv ist, wird
die `defaultString` Methode geladen, wenn das Profil `test` oder `test2` aktiv ist, wird
die `testString` Methode geladen, wenn keines der obengenannten Profile aktiv ist, wird

> die `rewoltString` Methode geladen.

Author: ```suggestion
die `rewoltString`-Methode geladen.
```
